--- /dev/null   2011-07-13 09:58:33.573375248 +0200
+++ b/HS 2011-07-21 13:43:51.000000000 +0200
@@ -0,0 +1 @@
+Copyright 2011 Efficient Cloud Ltd, All rights reserved.
--- a/src/core/nginx.c
+++ b/src/core/nginx.c
@@ -213,6 +213,8 @@ main(int argc, char *const *argv)
 
     if (ngx_show_version) {
         ngx_log_stderr(0, "nginx version: " NGINX_VER);
+        ngx_log_stderr(0, "Copyright (C) 2002-2011 Igor Sysoev");
+        ngx_log_stderr(0, "Copyright (C) 2011 EfficientCloud Ltd. All rights reserved.");
 
         if (ngx_show_help) {
             ngx_log_stderr(0,

